yt/visualization/accumulators.py

```diff
@@ -0,0 +1,290 @@
+        path segment as well as the endpoint for the last segment
+
+    Returns
+    -------
+    accum : YTArray
+        The cumulative value of the field integral at each path
+        segment
+    """
+    # https://en.wikipedia.org/wiki/Line_integral
+    # np.linalg.norm returns a ndarray, so when multiplied by field_vals, which
+    # is a YTArray, this leads to incorrect units. There does not appear to be
+    # a unyt implementation of norm, as far as I'm aware, so units will be
+    # handled manually for the time being
+    accum = np.cumsum(field_vals[:-1].d * np.linalg.norm(p[1:].d - p[:-1].d, axis=1))
+    accum = YTArray(accum, field_vals.units * p.units)
+    return accum
+
+
+class Accumulators:
+    r"""
+    Container for creating and storing the path integrals of various
+    fields along various paths.
+
+    The class takes in a list of user-defined paths and a dataset. From
+    these, the user can compute the integral of any field in the dataset
+    along the given paths. The results of these path integrals are
+    stored in a cumulative fashion, which means that, once computed, the
+    accumulated value of the field is stored at each point along the
+    path, allowing the user to query this information.
+    """
+
+    def __init__(self, paths, ds):
+        self.paths = []
+        self.ds = ds
+        self.ad = ds.all_data()
+        self.accum = []
+        self.left_edge = self.ds.domain_left_edge
+        self.right_edge = self.ds.domain_right_edge
+        # Make sure that the path has proper units. If no units are
+        # specified, assume code units
+        for p in paths:
+            if not isinstance(p, YTArray) or p.units.is_dimensionless:
+                self.paths.append(self.ds.arr(p, "code_length"))
+            else:
+                self.paths.append(p.to(self.ds.length_unit))
+
+    def _get_tree(self, field, is_vector):
+        r"""
+        Creates an AMRKDTree and then adds the field to it.
+
+        Parameters
+        ----------
+        field : list
+            Either the scalar field to add to the tree or an iterable
+            containing the components of the vector field to add.
+
+        is_vector : bool
+            If True, then `field` is a vector and it is assumed that it's
+            an iterable containing the x, y, and z components of the field.
+            Otherwise, the field is assumed to be a scalar field.
+
+        Returns
+        -------
+        tree : AMRKDTree
+            An AMRKDTree for the dataset containing the given field values.
+        """
+        tree = AMRKDTree(self.ds)
+        use_log = [False for i in range(len(field))]
+        field = [self.ad._determine_fields(f)[0] for f in field]
+        tree.set_fields(field, use_log, False)
+        return tree
+
+    def _get_path_field_values(self, tree, path, idx, vals, npts):
+        r"""
+        Determines the value of the field at each point along the path.
+
+        Parameters
```

Comment:
these parameters are out of date -- what are `idx` `vals` `npts`?  I'm not sure I understand this.